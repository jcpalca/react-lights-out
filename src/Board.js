import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows = 5, ncols = 5, chanceLightStartsOn = .25 }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [...Array(nrows)].map(
      r => [...Array(ncols)].map(
        (c) => Math.random() < chanceLightStartsOn
          ? true
          : false));

    // TODO: create array-of-arrays of true/false values
    return initialBoard;
  }

  function hasWon() {
    // TODO: check the board in state to determine whether the player has won.
  }


  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      // TODO: Make a (deep) copy of the oldBoard

      // TODO: in the copy, flip this cell and the cells around it

      // TODO: return the copy
    });
  }

  // if the game is won, just show a winning msg & render nothing else

  // TODO

  // make table board

  // for each item in row, for each item in col => render cell(flipCellsAroundMe, isLit )

  return (
    <div className="board">
      {
        board.map(
          (row, i) =>
            <div> {row.map(
              (cell, j) => <Cell key={i + '-' + j} isLit={board[i][j]} flipCellsAroundMe={() => flipCellsAround(i + '-' + j)} />)}

            </div>)
      }

    </div>);

  // TODO
}

export default Board;
