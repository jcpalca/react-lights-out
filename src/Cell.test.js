import { render } from "@testing-library/react";
import Cell from "./Cell.js";

import { board } from './Board.js';


describe("Renders Cell", function () {

  it("renders without crashing", function () {
    let container;
    let tr = document.createElement("tr");
    container = document.body.appendChild(tr);
    render(<Cell />, { container });
  });

  it("matches snapshot", function () {
    const { container } = render(<Cell isLit={true} />);
    expect(container).toMatchSnapshot();

  });

});
