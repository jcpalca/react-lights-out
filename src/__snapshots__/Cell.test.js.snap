// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Cell matches snapshot 1`] = `
<div>
  <td
    class="Cell Cell-lit"
  />
</div>
`;
